### PR TITLE
[Traceable FSDP2] Preserve fsdp.set_ op through lowering; Add unit test for multiple .set_ into same primal; Add unit test for FSDP2 module layer reuse

### PR DESCRIPTION
**Pull request by yf225**
Test commands:
- `pytest -rA test/distributed/_composable/fsdp/test_fully_shard_compile.py::TestFullyShardCompile::test_nested_fully_shard_fullgraph_backend_inductor`
- `pytest -rA test/functorch/test_aotdispatch.py::TestAOTAutograd::test_input_mutation_fsdp_set__into_same_input`
- `PYTORCH_TEST_WITH_CROSSREF=1 python test/functorch/test_aotdispatch.py -k TestAOTAutogradWithCache.test_input_mutation_fsdp_set__into_same_input`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130786
* #129773

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/fx/node.py [hunk 1]
```diff
@@ -650,7 +650,11 @@ def is_impure(self):
 
         # Check if an impure function.
         if self.op == "call_function":
-            return self.target in _side_effectful_functions
+            if self.target in _side_effectful_functions:
+                return True
+            if isinstance(self.target, torch._ops.OpOverload) and self.target._schema.is_mutable:
```
Comment (bdhirsh) [inline torch/fx/node.py:655]: hmm this feels right, but... i wonder why we haven't just always set up FX's impure check to check schema. Maybe @angelayi or @zou3519 if you happen to know?

File: test/functorch/test_aotdispatch.py [hunk 2]
```diff
@@ -724,6 +724,50 @@ def forward(self, primals_1):
...
+            z = a * a
+            return x + y + z
+
+        inp = [torch.ones(3, 3, requires_grad=True)]
+        fw_graph = self.verify_aot_autograd(
+            f, inp, test_mutation=True, keep_inp_mutations=True
+        )
+        inp = [torch.ones(3, 3, requires_grad=False)]
+        self.verify_aot_autograd(f, inp, test_mutation=True, keep_inp_mutations=True)
+        self.assertExpectedInline(
```
Comment (bdhirsh) [inline test/functorch/test_aotdispatch.py:751]: do you mind just mentioning the context somewhere (e.g. comment in this test):

(1) the important bit is that when there are multiple set_() calls on a graph input, they all show up with a primal as the first arg in the graph

(2) this doesn't actually work today with normall aten.set_, blocked on https://github.com/pytorch/pytorch/issues/129892, but using a custom op in fsdp with no returns is a simple workaround.

File: torch/_functorch/_aot_autograd/functional_utils.py [hunk 3]
```diff
@@ -406,16 +406,16 @@ def assert_functional_graph(fx_g: torch.fx.Graph) -> int:
...
                 suffix = True
-                # Can only copy_/set_ into an input, and can only do so once
+                # Can only copy_/set_ into an input
                 # this is mostly a hack to avoid failing XLA tests.
                 # See https://github.com/pytorch/pytorch/pull/122434#issuecomment-2101012113
                 if "set_buffer_donor_" not in str(n.args[0]):
                     assert (
                         n.args[0] in placeholders
                     ), f"n={str(n)}, n.args[0]={str(n.args[0])}, placeholders={str(placeholders)}, graph={str(fx_g)}"
-                    placeholders.remove(n.args[0])
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/functional_utils.py:418]: oh yeah removing this assert... makes sense (we can now end up with multiple mutations on the same graph input.

**Review by bdhirsh — APPROVED** [review]
stamped. I think the change so `fx.node.is_impure()` has a slightly higher blast radius -  might be good to keep in a separate PR

File: torch/fx/node.py (hunk 1 above)
Comment (yushangdi) [reply torch/fx/node.py:655]: I'll take a look! I think we should be able to add the check in #130552 (it check the schema) to the default DCE pass. Previously some tests were failing, but it should have been fixed now by #130598.

Comment (yf225) [reply torch/fx/node.py:655]: sounds great! I will revert back this change in my PR and just do a point fix for the op we care about here. Would be great to check the schema in general :)

**Pull request closed without merging**